content/source/docs/plugin/framework/validation.html.md

```diff
@@ -0,0 +1,257 @@
+
+Any further validation provided by the framework occurs after these checks.
+
+## Attribute Validation
+
+It is common for provider implementations to introduce validation on attributes using the generic framework-defined types such as `types.String`. The [`tfsdk.Attribute` type `Validators` field](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#Attribute.Validators) can be supplied with a list of validations and diagnostics will be returned from all validators. For example:
+
+```go
+// Typically within the tfsdk.Schema returned by GetSchema() for a provider,
+// resource, or data source.
+tfsdk.Attribute{
+    // ... other Attribute configuration ...
+
+    Validators: []AttributeValidators{
+        // These are example validators
+        stringLengthBetween(10, 256),
+        stringRegularExpression(regexp.MustCompile(`^[a-z0-9]+$`)),
+    },
+}
+```
+
+-> The framework will implement or reference common use case attribute validations, such as string length, in the future.
+
+### Creating Attribute Validators
+
+To create an attribute validator, the [`tfsdk.AttributeValidator` interface](https://pkg.go.dev/github.com/hashicorp/terraform-plugin-framework/tfsdk#AttributeValidator) must be satisfied. For example:
+
+```go
+type stringLengthBetweenValidator struct {
+    Max int
+    Min int
+}
+
+func (v stringLengthBetweenValidator) Description(ctx context.Context) string {
+    return fmt.Sprintf("string length must be between %d and %d", v.Min, v.Max)
+}
+
+func (v stringLengthBetweenValidator) MarkdownDescription(ctx context.Context) string {
+    return fmt.Sprintf("string length must be between `%d` and `%d`", v.Min, v.Max)
+}
+
+func (v stringLengthBetweenValidator) Validate(ctx context.Context, req tfsdk.ValidateAttributeRequest, resp *tfsdk.ValidateAttributeResponse) {
+    str, ok := request.AttributeConfig.(types.String)
+
+    if !ok {
+        response.Diagnostics.AddAttributeError(
+            request.AttributePath,
+            "Invalid Attribute Value Type",
+            fmt.Sprintf("Expected types.String value, got: %T", request.AttributeConfig),
+        )
+
+        return
+    }
+
+    if str.Unknown || str.Null {
+        return
+    }
+
+    strLen := len(str)
+
+    if strLen < v.Min || strLen > v.Max {
+        response.Diagnostics.AddAttributeError(
+            request.AttributePath,
+            "Attribute Validation Error",
```

Comment:
Can we be more descriptive here, to prompt provider devs to write better errors?

```suggestion
            "Invalid String Length",
```